### PR TITLE
`Stacked` bijector for stacking together bijectors

### PR DESCRIPTION
**Pull request by torfjelde**
This PR introduces `Stacked <: Bijector` which allows the user to, well, stack univariate and multivariate bijectors together into a single multivariate bijector. I've been using this a lot in a local branch of mine, both for experimentation with Bijectors.jl and improving VI in Turing.jl.

The concept of `Stacked` is fairly simple: the user doesn't have to write `[bijectors[i](x[ranges[i]]) for i = 1:length(ranges)]` to apply different bijectors across a vector where different parts of the vector require different bijectors.

But more than taking this tiny burden off the shoulders of the user, we can ensure that other computations for such a bijector is done correctly and efficiently, and that it works nicely with the rest of Bijectors.jl (e.g. composition). For example, the jacobian of such a `Stacked` is simply a block-diagonal matrix, where the i-th block is `jacobian(bijectors[i], x[ranges[i]])`. This also means that the `logabsdetjac(stacked, x)` is just `sum([logabsdetjac(bijectors[i], x[ranges[i]]) for i = ...])`. Moreover, we can also provide these implementations in a type-stable manner using `@generated` and whatnot.

Similarily to `Composed`, `Stacked` defaults to using a `Tuple` as the internal representation of the stacked bijectors, but allow using other types of containers.

I think the usefulness is most easily seen through some examples.

# Examples
## Simple
```julia
julia> using Bijectors; using Bijectors: Logit, Exp

julia> b1 = inv(Logit(0.0, 1.0))  # image is (0, 1)
Inversed{Logit{Float64}}(Logit{Float64}(0.0, 1.0))

julia> b2 = Bijectors.Exp()       # image is (0, ∞)
Exp()

julia> sb = vcat(b1, b2)  # or equivalently Stacked((b1, b2))
Stacked{Tuple{Inversed{Logit{Float64}},Exp},2}((Inversed{Logit{Float64}}(Logit{Float64}(0.0, 1.0)), Exp()), (1:1, 2:2))

julia> x = [-1.0, -3.0]
2-element Array{Float64,1}:
 -1.0
 -3.0

julia> y = sb(x)
2-element Array{Float64,1}:
 0.2689414213699951  
 0.049787068367863944

julia> (0 ≤ y[1] ≤ 1.0) && (0 < y[2])
true
```
## ADVI with `MvNormal` as base distribution
This `Stacked` allows us to do neat things like take a multivariate distribution as the base-distribution, which is then transformed using bijectors. This is very useful for a mean-field ADVI approach!
```julia
# Original distributions
dists = [
    Beta(),
    InverseGamma(),
    Dirichlet(2, 3)
]

# Construct the corresponding ranges
ranges = []
idx = 1
for i = 1:length(dists)
    d = dists[i]
    push!(ranges, idx:idx + length(d) - 1)

    global idx
    idx += length(d)
end

# Base distribution; mean-field normal
num_params = ranges[end][end]
d = MvNormal(zeros(num_params), ones(num_params))

# Construct the transform
bs = bijector.(dists)     # constrained-to-unconstrained bijectors for dists
ibs = inv.(bs)            # invert, so we get unconstrained-to-constrained
sb = Stacked(ibs, ranges) # => Stacked <: Bijector

# Mean-field normal with unconstrained-to-constrained stacked bijector
td = transformed(d, sb)
y = rand(td)

0.0 ≤ y[1] ≤ 1.0   # => true
0.0 < y[2]         # => true
sum(y[3:4]) ≈ 1.0  # => true
```
This can easily replace the hard-coded `elbo` in the current ADVI implementation in Turing. I have a local branch of Turing for which I have done this, and the resulting implementation is much more general. The only remaining part that needs to be addressed in that case is dealing with constraints on the parameters of the base distribution, e.g. ensuring variances of the underlying `MvNormal` are positive. When that is addressed (which can be done using the bijectors from Bijectors.jl!), we can just plug in arbitrary variational distributions to the `advi` call!

## Normalizing flows
With this `Stacked` bijector, it becomes trivial to create normalizing flows with constrained outputs!
```julia
using Bijectors, Turing, Tracker

# Simple model using Turing
@model gdemo_d() = begin
  s ~ InverseGamma(2, 3)
  m ~ Normal(0, sqrt(s))
  1.5 ~ Normal(m, sqrt(s))
  2.0 ~ Normal(m, sqrt(s))
  return s, m
end

# Computes the logdensity of a Turing.Model
logdensity(model, z) = begin
    varinfo = Turing.VarInfo(Turing.VarInfo(model), Turing.SampleFromUniform(), z)
    model(varinfo)
    return varinfo.logp
end

# Instantiate model
model = gdemo_d()

# Radial flow with an shift, composed with unconstrained-to-constrained transform
d = MvNormal(zeros(2), ones(2))
sb = vcat(bijector(InverseGamma(2, 3)), bijector(Normal()))
b = inv(sb) ∘ RadialLayer(2, param) ∘ Shift(param(zeros(2)))

# Sample from `d`, and transform using `b`
flow = transformed(d, b)

# Sample and transform from flow
z₀, zₖ, logjac, logpdf_q = forward(flow)

# Compute logdensity using the sampled parameters
lp = logdensity(model, zₖ)

# Since `q isa MvNormal` we have analytical expression for entropy
elbo = lp - logjac + entropy(q)

# Backprop the negative ELBO
Tracker.back!(- elbo, 1.0)

# Perform gradient descent using your favorite optimiser!
...
```
By using Flux's `treelike` and whatnot to traverse the Bijectors and their corresponding parameters, the update loop simply becomes
```julia
using Flux: Optimise
using Flux: treelike
@treelike Shift
@treelike RadialFlow
@treelike Stacked
@treelike TransformedDistribution

opt = Optimise.ADAGrad()

for p in Flux.params(flow)
    Δ = Tracker.grad(p)
    Optimise.update!(opt, p, Δ)
end
```
With the changes `Stacked` allows to ADVI in Turing, we can easily just plug the resulting posterior into `advi` to have NFVI + ADVI for (possibly) more flexible VI! It's all very exciting, isn't it?

# Things to address
- [x] Currently we provide the following different ways of constructing a `Stacked`. I think it's up for discussion whether we want the two "default" constructors `Stacked(bs)` and `vcat(bs...)`.
  - `Stacked(bs, ranges)` which is fairly straight forward
  - `Stacked(bs)` which assumes all the bijectors `bs` are _univariate_, i.e. the `ranges` in the original constructor is replaced by `[1:1, 2:2, 3:3, ...]`. Right now this makes it very easy for the user to do the _wrong_ thing, though in the transform performed by `Stacked` we have an `@assert size(x) == size(y)` where `x` is input and `y` is transformed to let the user know they did something wrong:) This can probably be caught upon construction when we've resolved https://github.com/TuringLang/Bijectors.jl/issues/35 in a good manner.
  - `vcat(bs::Bijector...) = Stacked(bs)`, suffers from same issues as `Stacked(bs)`.

### CODE REVIEW

File: src/interface.jl [hunk 1]
```diff
@@ -123,7 +124,9 @@ end
 struct SingularJacobianException{B} <: Exception where {B<:Bijector}
     b::B
 end
-Base.showerror(io::IO, e::SingularJacobianException) = print(io, "jacobian of $(e.b) is singular")
+function Base.showerror(io::IO, e::SingularJacobianException)
+    print(io, "jacobian of $(e.b) is singular")
+end
```
Comment (xukai92) [inline src/interface.jl]: Is this for line character limit?

File: src/interface.jl [hunk 2]
```diff
@@ -324,13 +401,14 @@ logabsdetjac(b::Scale, x) = log(abs(b.a))
 # Simplex bijector #
 ####################
 struct SimplexBijector{T} <: Bijector where {T} end
+SimplexBijector() = SimplexBijector{Val{true}}()
 
 const simplex_b = SimplexBijector{Val{false}}()
 const simplex_b_proj = SimplexBijector{Val{true}}()
```
Comment (xukai92) [inline src/interface.jl]: Is this still used?

File: src/interface.jl [hunk 3]
```diff
@@ -725,3 +811,26 @@ In the case where `d isa Distribution`, this means
 """
 forward(d::Distribution) = forward(GLOBAL_RNG, d)
 forward(d::Distribution, num_samples::Int) = forward(GLOBAL_RNG, d, num_samples)
+
+# utility stuff
+params(td::Transformed) = params(td.dist)
```
Comment (xukai92) [inline src/interface.jl]: Is `Transformed` treelike?

File: src/interface.jl [hunk 4]
```diff
@@ -240,6 +243,80 @@ function forward(cb::Composed, x)
...
+    y = _transform(x, sb.ranges, sb.bs...)
+
+    # TODO: maybe tell user to check their ranges?
+    @assert size(y) == size(x)
+
+    return y
+end
+(sb::Stacked)(x::AbstractMatrix{<: Real}) = hcat([sb(x[:, i]) for i = 1:size(x, 2)]...)
+function (sb::Stacked)(x::TrackedArray{A, 2}) where {A}
+    return Tracker.collect(hcat([sb(x[:, i]) for i = 1:size(x, 2)]...))
```
Comment (xukai92) [inline src/interface.jl]: Can you remind me of what does `Tracker.collect` do?

File: test/interface.jl [hunk 5]
```diff
@@ -282,6 +282,94 @@ struct NonInvertibleBijector{AD} <: ADBijector{AD} end
         @test f_t == f_a
     end
 
+    @testset "Stacked <: Bijector" begin
```
Comment (xukai92) [inline test/interface.jl]: Can we add a simple test that do something like below?
```julia
x = ones(5)
sb = vcat(Exp(), Log(), ...) # stack some simple bijectors
@test sb(x) == [...] # some simple exact answers
```
```

File: src/interface.jl (hunk 1 above)
Comment (xukai92) [reply src/interface.jl]: If so I suggest to do the following
```julia
Base.showerror(io::IO, e::SingularJacobianException) = 
    print(io, "jacobian of $(e.b) is singular")

```

File: src/interface.jl (hunk 4 above)
Comment (torfjelde) [reply src/interface.jl]: Basically convert `Array{TrackedReal}` into `TrackedArray`

File: src/interface.jl (hunk 2 above)
Comment (torfjelde) [reply src/interface.jl]: Not sure? I thought it was nice to have due to this basically corresponding to the default `link` and `invlink`.

File: src/interface.jl (hunk 3 above)
Comment (torfjelde) [reply src/interface.jl]: No, not yet at least. It's `params` from Distributions.jl but yeah it doesn't catch the case where the bijectors themselves are parameterized. It's not clear to me how to do this in general. There's a discussion regarding whether or not we want to depend on Flux over in Sharans PR on batch norm

File: src/interface.jl (hunk 1 above)
Comment (torfjelde) [reply src/interface.jl]: Yup!
But does that follow the [style-guide](https://turing.ml/dev/docs/contributing/style-guide#method-definitions): "Only use short-form function definitions when they fit on a single line"?

File: src/interface.jl (hunk 2 above)
Comment (torfjelde) [reply src/interface.jl]: Should maybe also include a `SimplexBijector(proj::bool) = SimplexBijector{Val{proj}}()`?

File: src/interface.jl (hunk 1 above)
Comment (xukai92) [reply src/interface.jl]: Opps. I think it is me who is not following the style guide.

Comment (xukai92) [reply src/interface.jl]: Let's stick to the guide then. Thanks for pointing out this.

File: src/interface.jl (hunk 2 above)
Comment (xukai92) [reply src/interface.jl]: Looks like a good constructor singature to me.

**Comment by torfjelde** [comment]
Btw, what do people think about the "Things to address" part of the PR?

Currently you can construct a `Stacked` using `vcat`; but this will assume they're all 0-dimensional bijectors. In #44 were we introduce the dimensionality into the type, we can specialize `vcat(bs::Bijector{0}...)` which ensures the correct thing is done. We can also do this for `Stacked(bs::Bijector{0}...)`.

File: src/interface.jl [hunk 6]
```diff
@@ -256,6 +259,80 @@ function forward(cb::Composed, x)
...
+end
+
+@generated function _logabsdetjac(
+    x,
+    rs::NTuple{N, UnitRange{Int}},
+    bs::Bijector...
+) where {N}
+    exprs = []
+    for i = 1:N
+        push!(exprs, :(sum(logabsdetjac(bs[$i], x[rs[$i]]))))
```
Comment (torfjelde) [inline src/interface.jl]: This `sum` here was to deal with the case when `rs[i] == 2:2` and `bs[i]` is a 0-dim bijector; in this case then the bijector is only guaranteed to be defined for a `Real` so we should ideally pass it `x[i]` if `length(rs[i]) == 1`. Does that make sense?

This is another issue (also the `vcat` one) solved by #44

**Comment by xukai92** [comment]
> Btw, what do people think about the "Things to address" part of the PR?
> 
> Currently you can construct a `Stacked` using `vcat`; but this will assume they're all 0-dimensional bijectors. In #44 were we introduce the dimensionality into the type, we can specialize `vcat(bs::Bijector{0}...)` which ensures the correct thing is done. We can also do this for `Stacked(bs::Bijector{0}...)`.
> 
> But with current master it's very easy for the user to do the _wrong_ thing :/

It's OK to leave it as it as #44 will come soon.

Also, how about`vcat(bs::Bijector{1}...)`?

**Comment by torfjelde** [comment]
> Also, how about `vcat(bs::Bijector{1}...)`?
The issue is that we can't just deduce the corresponding `ranges`. In the case where we have multivariate bijectors, the user needs to explicitly pass in the ranges using `Stacked(bs, ranges)`

**Comment by xukai92** [comment]
I guess we can support `Stacked(bs, ranges)` and `Stacked(bs::StaticBijector{1}...)` then.

**Comment by torfjelde** [comment]
> I guess we can support Stacked(bs, ranges) and Stacked(bs::StaticBijector{1}...) then.

Yup!:)

File: src/interface.jl [hunk 7]
```diff
@@ -434,6 +434,7 @@ end
 
 function (b::SimplexBijector{Val{proj}})(x::AbstractVector{T}) where {T, proj}
     y, K = similar(x), length(x)
+    @assert K > 1
```
Comment (xukai92) [inline src/interface.jl]: Can you add assertion msgs as well?

File: test/interface.jl (hunk 5 above)
Comment (xukai92) [reply test/interface.jl]: Is this done?

Comment (torfjelde) [reply test/interface.jl]: Yup: L339

File: test/interface.jl [hunk 8]
```diff
@@ -282,6 +282,94 @@ struct NonInvertibleBijector{AD} <: ADBijector{AD} end
...
+            bs = bijector.(dists)     # constrained-to-unconstrained bijectors for dists
+            ibs = inv.(bs)            # invert, so we get unconstrained-to-constrained
+            sb = Stacked(ibs, ranges) # => Stacked <: Bijector
+            x = rand(d)
+            sb(x)
+            @test sb isa Stacked
+
+            td = transformed(d, sb)  # => MultivariateTransformed <: Distribution{Multivariate, Continuous}
+            @test td isa Distribution{Multivariate, Continuous}
+
```
Comment (xukai92) [inline test/interface.jl]: L339 looks like an emtpy line.

Comment (torfjelde) [reply test/interface.jl]: Yes, I've just added some whitespace to indicate a separation between the two tests, this and the one that "starts" at L341

File: src/interface.jl (hunk 2 above)
Comment (torfjelde) [reply src/interface.jl]: I've also removed the `simplex_b` and `simplex_b_proj` as they are no longer used (Maybe this was actually what you were asking but for some reason Github showed the `SimplexBijector()` line as the one you were referring to)

**Comment by xukai92** [comment]
24-hour merge count down.

**Comment by xukai92** [comment]
In the meantime can you bump up the version number - I will make a new release after this PR is merged.

**Comment by torfjelde** [comment]
I'd suggest one final change: rename `vcat` to `stack`. I've realized it wasn't such a good idea to overload `Base.vcat` since this means we can't actually construct arrays of bijectors using `vcat`. I initially got a bit carried away with making it "intuitive" to just stack bijectors :confused:

**Pull request merged**